File: Plugins/AM_Mz_Refinery_Plugin/clsAnalysisResourcesMzRefinery.vb
Option Strict On

Imports AnalysisManagerBase
Imports System.IO

Public Class clsAnalysisResourcesMzRefinery
    Inherits clsAnalysisResources

#Region "Methods"

    Public Overrides Sub Setup(mgrParams As IMgrParams, jobParams As IJobParams, statusTools As IStatusFile, myEMSLUtilities As clsMyEMSLUtilities)
        MyBase.Setup(mgrParams, jobParams, statusTools, myEmslUtilities)
        SetOption(clsGlobal.eAnalysisResourceOptions.OrgDbRequired, True)
    End Sub

    ''' <summary>
    ''' Retrieves files necessary for running MzRefinery
    ''' </summary>
    ''' <returns>IJobParams.CloseOutType indicating success or failure</returns>
    ''' <remarks></remarks>
    Public Overrides Function GetResources() As IJobParams.CloseOutType

        Dim currentTask = "Initializing"

        Try

            currentTask = "Retrieve shared resources"

            ' Retrieve shared resources, including the JobParameters file from the previous job step
            GetSharedResources()

            Dim mzRefParamFile = m_jobParams.GetJobParameter("MzRefParamFile", String.Empty)
            If String.IsNullOrEmpty(mzRefParamFile) Then
                LogError("MzRefParamFile parameter is empty")
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            currentTask = "Get Input file"

            Dim eResult As IJobParams.CloseOutType
            eResult = GetMsXmlFile()

            If eResult <> IJobParams.CloseOutType.CLOSEOUT_SUCCESS Then
                Return eResult
            End If

            ' Retrieve the Fasta file
            Dim localOrgDbFolder = m_mgrParams.GetParam("orgdbdir")

            currentTask = "RetrieveOrgDB to " & localOrgDbFolder

            If Not RetrieveOrgDB(localOrgDbFolder) Then
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            ' Retrieve the Mz Refinery parameter file
            currentTask = "Retrieve the Mz Refinery parameter file " & mzRefParamFile

            Const paramFileStoragePathKeyName As String = clsGlobal.STEPTOOL_PARAMFILESTORAGEPATH_PREFIX & "Mz_Refinery"

            Dim mzRefineryParmFileStoragePath = m_mgrParams.GetParam(paramFileStoragePathKeyName)
            If String.IsNullOrWhiteSpace(mzRefineryParmFileStoragePath) Then
                mzRefineryParmFileStoragePath = "\\gigasax\dms_parameter_Files\MzRefinery"
                clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Parameter '" & paramFileStoragePathKeyName & "' is not defined (obtained using V_Pipeline_Step_Tools_Detail_Report in the Broker DB); will assume: " & mzRefineryParmFileStoragePath)
            End If

            If Not RetrieveFile(mzRefParamFile, mzRefineryParmFileStoragePath) Then
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

            ' Look for existing MSGF+ results in the transfer folder
            currentTask = "Find existing MSGF+ results"

            If Not FindExistingMSGFPlusResults(mzRefParamFile) Then
                Return IJobParams.CloseOutType.CLOSEOUT_FAILED
            End If

        Catch ex As Exception
            m_message = "Exception in GetResources: " & ex.Message
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message & "; task = " & currentTask & "; " & clsGlobal.GetExceptionStackTrace(ex))
            Return IJobParams.CloseOutType.CLOSEOUT_FAILED
        End Try

        Return IJobParams.CloseOutType.CLOSEOUT_SUCCESS

    End Function

    ''' <summary>
    ''' Check for existing MSGF+ results in the transfer directory
    ''' </summary>
    ''' <returns>True if no errors, false if a problem</returns>
    ''' <remarks>Will retrun True even if existing results are not found</remarks>
    Private Function FindExistingMSGFPlusResults(mzRefParamFileName As String) As Boolean

        Dim resultsFolderName = m_jobParams.GetParam("OutputFolderName")
        Dim transferFolderPath = m_jobParams.GetParam("transferFolderPath")

        If String.IsNullOrWhiteSpace(resultsFolderName) Then
            m_message = "Results folder not defined (job parameter OutputFolderName)"
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
            Return False
        End If

        If String.IsNullOrWhiteSpace(transferFolderPath) Then
            m_message = "Transfer folder not defined (job parameter transferFolderPath)"
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.ERROR, m_message)
            Return False
        End If

        Dim diTransferFolder = New DirectoryInfo(Path.Combine(transferFolderPath, m_DatasetName, resultsFolderName))
        If Not diTransferFolder.Exists Then
            ' This is not an error -- it just means there are no existing MSGF+ results to use
            Return True
        End If

        ' Look for the required files in the transfer folder
        Dim resultsFileName = m_DatasetName & clsAnalysisToolRunnerMzRefinery.MSGFPLUS_MZID_SUFFIX & ".gz"
        Dim fiMSGFPlusResults = New FileInfo(Path.Combine(diTransferFolder.FullName, resultsFileName))

        If Not fiMSGFPlusResults.Exists Then
            ' This is not an error -- it just means there are no existing MSGF+ results to use
            Return True
        End If

        Dim fiMSGFPlusConsoleOutput = New FileInfo(Path.Combine(diTransferFolder.FullName, "MSGFPlus_ConsoleOutput.txt"))
        If Not fiMSGFPlusResults.Exists Then
            ' This is unusual; typically if the mzid.gz file exists there should be a ConsoleOutput file
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Found " & fiMSGFPlusResults.FullName & " but did not find " & fiMSGFPlusConsoleOutput.Name & "; will re-run MSGF+")
            Return True
        End If

        Dim fiMzRefParamFile = New FileInfo(Path.Combine(diTransferFolder.FullName, mzRefParamFileName))
        If Not fiMzRefParamFile.Exists Then
            ' This is unusual; typically if the mzid.gz file exists there should be a MzRefinery parameter file
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.WARN, "Found " & fiMSGFPlusResults.FullName & " but did not find " & fiMzRefParamFile.Name & "; will re-run MSGF+")
            Return True
        End If

        ' Compare the remote parameter file and the local one to make sure they match
        If Not clsGlobal.TextFilesMatch(fiMzRefParamFile.FullName, Path.Combine(m_WorkingDir, mzRefParamFileName), True) Then
            clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, "MzRefinery parameter file in transfer folder does not match the official MzRefinery paramter file; will re-run MSGF+")
            Return True
        End If

        ' Existing results found
        ' Copy the MSGF+ results locally
        Dim localFilePath = Path.Combine(m_WorkingDir, fiMSGFPlusResults.Name)
        fiMSGFPlusResults.CopyTo(localFilePath, True)

        m_IonicZipTools.GUnzipFile(localFilePath)

        localFilePath = Path.Combine(m_WorkingDir, fiMSGFPlusConsoleOutput.Name)
        fiMSGFPlusConsoleOutput.CopyTo(localFilePath, True)

        clsLogTools.WriteLog(clsLogTools.LoggerTypes.LogFile, clsLogTools.LogLevels.INFO, "Found existing MSGF+ results to use for MzRefinery")

        Return True

    End Function

#End Region

End Class

File: Plugins/AM_Mz_Refinery_Plugin/clsAnalysisResourcesMzRefinery.cs
using System;

using AnalysisManagerBase;
using System.IO;

namespace AnalysisManagerMzRefineryPlugIn
{

    /// <summary>
    /// Retrieve resources for the MzRefinery plugin
    /// </summary>
    public class clsAnalysisResourcesMzRefinery : clsAnalysisResources
    {

        /// <summary>
        /// Initialize options
        /// </summary>
        public override void Setup(string stepToolName, IMgrParams mgrParams, IJobParams jobParams, IStatusFile statusTools, clsMyEMSLUtilities myEMSLUtilities)
        {
            base.Setup(stepToolName, mgrParams, jobParams, statusTools, myEMSLUtilities);
            SetOption(clsGlobal.eAnalysisResourceOptions.OrgDbRequired, true);
        }

        /// <summary>
        /// Retrieves files necessary for running MzRefinery
        /// </summary>
        /// <returns>CloseOutType indicating success or failure</returns>
        /// <remarks></remarks>
        public override CloseOutType GetResources()
        {
            var currentTask = "Initializing";

            try
            {
                currentTask = "Retrieve shared resources";

                // Retrieve shared resources, including the JobParameters file from the previous job step
                GetSharedResources();

                var mzRefParamFile = m_jobParams.GetJobParameter("MzRefParamFile", string.Empty);
                if (string.IsNullOrEmpty(mzRefParamFile))
                {
                    LogError("MzRefParamFile parameter is empty");
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                currentTask = "Get Input file";

                var eResult = GetMsXmlFile();

                if (eResult != CloseOutType.CLOSEOUT_SUCCESS)
                {
                    return eResult;
                }

                // Retrieve the Fasta file
                var localOrgDbFolder = m_mgrParams.GetParam("orgdbdir");

                currentTask = "RetrieveOrgDB to " + localOrgDbFolder;

                if (!RetrieveOrgDB(localOrgDbFolder))
                {
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                // Retrieve the Mz Refinery parameter file
                currentTask = "Retrieve the Mz Refinery parameter file " + mzRefParamFile;

                const string paramFileStoragePathKeyName = clsGlobal.STEPTOOL_PARAMFILESTORAGEPATH_PREFIX + "Mz_Refinery";

                var mzRefineryParmFileStoragePath = m_mgrParams.GetParam(paramFileStoragePathKeyName);
                if (string.IsNullOrWhiteSpace(mzRefineryParmFileStoragePath))
                {
                    mzRefineryParmFileStoragePath = @"\\gigasax\dms_parameter_Files\MzRefinery";
                    LogWarning("Parameter '" + paramFileStoragePathKeyName + "' is not defined " +
                        "(obtained using V_Pipeline_Step_Tools_Detail_Report in the Broker DB); " +
                        "will assume: " + mzRefineryParmFileStoragePath);
                }

                if (!FileSearch.RetrieveFile(mzRefParamFile, mzRefineryParmFileStoragePath))
                {
                    return CloseOutType.CLOSEOUT_FAILED;
                }

                // Look for existing MSGF+ results in the transfer folder
                currentTask = "Find existing MSGF+ results";

                if (!FindExistingMSGFPlusResults(mzRefParamFile))
                {
                    return CloseOutType.CLOSEOUT_FAILED;
                }
            }
            catch (Exception ex)
            {
                m_message = "Exception in GetResources: " + ex.Message;
                LogError(m_message + "; task = " + currentTask + "; " + clsGlobal.GetExceptionStackTrace(ex));
                return CloseOutType.CLOSEOUT_FAILED;
            }

            return CloseOutType.CLOSEOUT_SUCCESS;
        }

        /// <summary>
        /// Check for existing MSGF+ results in the transfer directory
        /// </summary>
        /// <returns>True if no errors, false if a problem</returns>
        /// <remarks>Will retrun True even if existing results are not found</remarks>
        private bool FindExistingMSGFPlusResults(string mzRefParamFileName)
        {
            var resultsFolderName = m_jobParams.GetParam(JOB_PARAM_OUTPUT_FOLDER_NAME);
            var transferFolderPath = m_jobParams.GetParam(JOB_PARAM_TRANSFER_FOLDER_PATH);

            if (string.IsNullOrWhiteSpace(resultsFolderName))
            {
                m_message = "Results folder not defined (job parameter OutputFolderName)";
                LogError(m_message);
                return false;
            }

            if (string.IsNullOrWhiteSpace(transferFolderPath))
            {
                m_message = "Transfer folder not defined (job parameter transferFolderPath)";
                LogError(m_message);
                return false;
            }

            var diTransferFolder = new DirectoryInfo(Path.Combine(transferFolderPath, DatasetName, resultsFolderName));
            if (!diTransferFolder.Exists)
            {
                // This is not an error -- it just means there are no existing MSGF+ results to use
                return true;
            }

            // Look for the required files in the transfer folder
            var resultsFileName = DatasetName + clsAnalysisToolRunnerMzRefinery.MSGFPLUS_MZID_SUFFIX + ".gz";
            var fiMSGFPlusResults = new FileInfo(Path.Combine(diTransferFolder.FullName, resultsFileName));

            if (!fiMSGFPlusResults.Exists)
            {
                // This is not an error -- it just means there are no existing MSGF+ results to use
                return true;
            }

            var fiMSGFPlusConsoleOutput = new FileInfo(Path.Combine(diTransferFolder.FullName, "MSGFPlus_ConsoleOutput.txt"));
            if (!fiMSGFPlusResults.Exists)
            {
                // This is unusual; typically if the mzid.gz file exists there should be a ConsoleOutput file
                LogWarning("Found " + fiMSGFPlusResults.FullName + " but did not find " + fiMSGFPlusConsoleOutput.Name + "; will re-run MSGF+");
                return true;
            }

            var fiMzRefParamFile = new FileInfo(Path.Combine(diTransferFolder.FullName, mzRefParamFileName));
            if (!fiMzRefParamFile.Exists)
            {
                // This is unusual; typically if the mzid.gz file exists there should be a MzRefinery parameter file
                LogWarning("Found " + fiMSGFPlusResults.FullName + " but did not find " + fiMzRefParamFile.Name + "; will re-run MSGF+");
                return true;
            }

            // Compare the remote parameter file and the local one to make sure they match
            if (!clsGlobal.TextFilesMatch(fiMzRefParamFile.FullName, Path.Combine(m_WorkingDir, mzRefParamFileName), true))
            {
                LogMessage("MzRefinery parameter file in transfer folder does not match the official MzRefinery paramter file; will re-run MSGF+");
                return true;
            }

            // Existing results found
            // Copy the MSGF+ results locally
            var localFilePath = Path.Combine(m_WorkingDir, fiMSGFPlusResults.Name);
            fiMSGFPlusResults.CopyTo(localFilePath, true);

            GUnzipFile(localFilePath);

            localFilePath = Path.Combine(m_WorkingDir, fiMSGFPlusConsoleOutput.Name);
            fiMSGFPlusConsoleOutput.CopyTo(localFilePath, true);

            LogMessage("Found existing MSGF+ results to use for MzRefinery");

            return true;
        }

    }
}
